`C-cmse-nonsecure-call`: improved error messages

folkertdev opened the pull request:
tracking issue: #81391 
issue for the error messages (partially implemented by this PR): #81347 
related, in that it also deals with CMSE: https://github.com/rust-lang/rust/pull/127766

When using the `C-cmse-nonsecure-call` ABI, both the arguments and return value must be passed via registers. Previously, when violating this constraint, an ugly LLVM error would be shown. Now, the rust compiler itself will print a pretty message and link to more information.

folkertdev committed:
add rust error message for CMSE stack spill

when the `C-cmse-nonsecure-call` ABI is used, arguments and return values must be passed via registers. Failing to do so (i.e. spilling to the stack) causes an LLVM error down the line, but now rustc will properly emit an error a bit earlier in the chain

folkertdev committed:
add more tests for `cmse-nonsecure-call` stack spills

GuillaumeGomez reviewed (approved):
Diagnostic error explanation looks good to me.

oli-obk commented:
Can you swap the commits so we can see the diagnostics as a diff on your actual commit? Or are the LLVM diagnostics too annoying?

tdittr commented:
> Can you swap the commits so we can see the diagnostics as a diff on your actual commit? Or are the LLVM diagnostics too annoying?

We did consider this, but LLVM will only return the first error it encounters. Also rebasing this is not straight forward, since we moved some of the tests which `git rebase` really does not seem to like.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9959734748/job/27517270213) [(plain)](https://github.com/rust-lang/rust/commit/1e8606408d14e2509f59c0a87c02788cb112b55a/checks/27517270213/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.75s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
 finished in 35.450 seconds
##[group]Testing stage2 error-index (x86_64-unknown-linux-gnu)
##[endgroup]

Command CFG_RELEASE_CHANNEL="nightly" RUSTC_BOOTSTRAP="1" RUSTC_STAGE="2" RUSTC_SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" RUSTDOC_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTDOC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" RUST_TEST_THREADS="16" "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" "" (failure_mode=DelayFail, stdout_mode=Capture, stderr_mode=Capture) did not execute successfully.
Created at: src/core/builder.rs:1314:23
Executed at: src/core/build_steps/test.rs:2381:9

STDOUT ----
```

</details>

folkertdev committed:
make function pub in error_codes markdown file

the error is only generated for functions that are actually codegen'd

folkertdev committed:
move cmse ABI validation into its own module

Co-authored-by: Tamme Dittrich <tamme@tweedegolf.com>

tdittr commented:
For the name of the new mod we were not sure, so we just picked one, suggestions are very welcome.

folkertdev commented on compiler/rustc_codegen_ssa/src/mir/cmse.rs:
```diff
@@ -0,0 +1,72 @@
+use crate::errors::{CmseCallInputsStackSpill, CmseCallOutputStackSpill};
+use crate::traits::BuilderMethods;
+
+/// Check conditions on inputs and outputs that the cmse ABIs impose: arguments and results MUST be
+/// returned via registers (i.e. MUST NOT spill to the stack). LLVM will also validate these
+/// conditions, but by checking them here rustc can emit nicer error messages.
+pub fn validate_cmse_abi<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
```
in the near future, this will also validate the `C-cmse-nonsecure-entry` abi which will be added by https://github.com/rust-lang/rust/pull/127766.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9960851293/job/27521042900) [(plain)](https://github.com/rust-lang/rust/commit/c7ff46c971ea7c17b67b2b619fd094c4df981863/checks/27521042900/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.91s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
 finished in 35.555 seconds
##[group]Testing stage2 error-index (x86_64-unknown-linux-gnu)
##[endgroup]

Command CFG_RELEASE_CHANNEL="nightly" RUSTC_BOOTSTRAP="1" RUSTC_STAGE="2" RUSTC_SYSROOT="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" RUSTDOC_LIBDIR="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" RUSTDOC_REAL="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" RUST_TEST_THREADS="16" "/checkout/obj/build/bootstrap/debug/rustdoc" "-Wrustdoc::invalid_codeblock_attributes" "-Dwarnings" "-Znormalize-docs" "-Z" "unstable-options" "--test" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md" "--test-args" "" (failure_mode=DelayFail, stdout_mode=Capture, stderr_mode=Capture) did not execute successfully.
Created at: src/core/builder.rs:1314:23
Executed at: src/core/build_steps/test.rs:2381:9

STDOUT ----
```

</details>

folkertdev commented:
CI fails with

```
failures:

---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0798 (line 18059) stdout ----
Test compiled successfully, but it's marked `compile_fail`.
---- /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0798 (line 18073) stdout ----
Test compiled successfully, but it's marked `compile_fail`.

failures:
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0798 (line 18059)
    /checkout/obj/build/x86_64-unknown-linux-gnu/test/error-index.md - Rust_Compiler_Error_Index::E0798 (line 18073)

test result: FAILED. 1041 passed; 2 failed; 55 ignored; 0 measured; 0 filtered out; finished in 8.37s
```

How can I run those tests locally?

folkertdev committed:
another attempt at fixing the examples in the error codes .md

folkertdev commented on compiler/rustc_error_codes/src/error_codes/E0798.md:
````diff
@@ -0,0 +1,38 @@
+
+Erroneous code example:
+
+```compile_fail,E0798
+#![feature(abi_c_cmse_nonsecure_call)]
+
+#[no_mangle]
+pub fn test(
+    f: extern "C-cmse-nonsecure-call" fn(u32, u32, u32, u32, u32) -> u32,
+) -> u32 {
+    f(1, 2, 3, 4, 5)
+}
+```
````
CI fails for the code samples in this file. I suspect something weird is going on because this snippet only gives the asserted error with a very specific target, and the test run on these files does not take the target into consideration.

@GuillaumeGomez are there similar examples among the other error codes? I'm not getting a clear error message and therefore not sure what to do here.

GuillaumeGomez replied:
Not that I know of.

folkertdev replied:
so, would it be ok to skip the "```compile_fail,E0798" bit here? then this doesn't really test anything anymore but that seems like it's the best I can do right now

GuillaumeGomez replied:
If it doesn't work on all target, I think it's ok.

folkertdev committed:
stop running code samples in the error code .md

folkertdev commented on tests/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack.stderr:
```diff
@@ -0,0 +1,58 @@
+error[E0798]: arguments for `C-cmse-nonsecure-call` function too large to pass via registers
+  --> $DIR/params-via-stack.rs:24:5
+   |
+LL |     f1: extern "C-cmse-nonsecure-call" fn(u32, u32, u32, u32, u32) -> u32,
```
we considered that but ultimately decided against that because 

- there is no risk of actual incorrect behavior (given the check at the call site)
- we don't know how to perform that check exhaustively

Is it feasible to disallow values of such types to ever exist? (e.g. in parameter lists, in types, as the output of `core::mem::transmute`).

We could consider adding lints later (e.g. an argument count of more than 4 is always incorrect).

oli-obk replied:
It could be done in `lower_fn_ty`. That's the one central place where function signatures are handled and we already have type information available, so you should be able to get the layout of the arguments and the return type.

folkertdev replied:
one thing I ran into earlier is that it's hard to unwrap `repr(transparent)` wrappers to get at the actual inner type. I was not able to do that earlier with just a `tcx` because it does not implement the required traits. In theory generics are also a problem I think though we could disallow them on these functions. Are there examples of where the `repr(transparant)` unwrapping  happens on HIR elsewhere? A point in favor of performing the check earlier would be that the error message shows up when running `cargo check` (currently it only shows when running `cargo build` or similar).

also in general, should we disallow a type that cannot be used (in any useful way) from being created (instead of just restricting its use). I can't really think of analogous cases either way.

oli-obk replied:
Yea we do try to reject types like this early and often just replace them with the error type after an error has been reported.

You don't need to process the HIR. `lower_fn_ty` converts the HIR to a `Ty<'tcx>` representation, so you can process it right after that happens.

folkertdev replied:
also to be clear: do you suggest to additionally error in `lower_fn_ty` if we can detect it, or to just try our best there, replace with some error type placeholder (no idea how to do that exactly but I'll have a look) and then still perform the check at codegen time if something slipped through the cracks? Or should all checking happen in `lower_fn_ty`?

oli-obk replied:
> that function just cannot be called in a non-monomorphic context.

you can just reject generic `cmse` ABI function signatures for now. The PR as is introduces new post-monomorphization errors that we usually try to avoid by adding appropriate trait bounds. Until you have a use for generic `csme` ABI functions, we should avoid them

folkertdev replied:
conveniently generics are not allowed on function pointers anyway, nor is `impl Trait`, but I'm struggling with cases like this 

```rust
pub fn test<T: Copy>(
    t: T,
    f: extern "C-cmse-nonsecure-call" fn(T, u32, u32, u32) -> u64,
) -> u64 {
    f(t, 2, 3, 4)
}
```

is it reliable to use `tcx.layout_of` returning `Err(_)` to indicate that a generic was used (and that hence the layout cannot be determined)? or is there some more idiomatic way to figure out whether there is any (nested) generic variable used in a type?

oli-obk replied:
Yea, relying on layout_of suffices

folkertdev committed:
add test for repr(transparent) enum

folkertdev committed:
move CMSE validation to hir_analysis

folkertdev committed:
add test for unions and repr(transparent) with ZST fields

folkertdev commented:
Something that stops working with the logic as implemented (and I think that is correct) is using a `repr(Rust)` union that is effectively a foundational type. 

https://godbolt.org/z/oaTdf9cYE

My reasoning is that `repr(Rust)` is unstable, and so while it is currently given to LLVM as a foundational type, that is not guaranteed. We're future-proofing the implementation here by not allowing it. Using a union with just one foundational field does not work in C. `repr(transparent)` on unions seems far away from being stabilized, but if it guarantees that a one-field union has the same representation as that field then that would work.

folkertdev committed:
remove cmse validation from rustc_codegen_ssa

it was moved to hir_analysis in the previous commit

folkertdev commented:
we've now implemented the logic in hir_analyze, removed code from codegen_ssa, and error when generics are used in function pointer types with the `C-cmse-nonsecure-call`  abi

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9989625404/job/27608657483) [(plain)](https://github.com/rust-lang/rust/commit/096cc198fe4e6b773509fafdd090023a7f516eab/checks/27608657483/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
1 error: function pointer types may not have generic parameters
-   --> $DIR/generics.rs:14:42
+   --> $DIR/generics.rs:15:42
3    |
4 LL |     f1: extern "C-cmse-nonsecure-call" fn<U: Copy>(U, u32, u32, u32) -> u64,

6 
7 error[E0412]: cannot find type `U` in this scope
-   --> $DIR/generics.rs:14:52
-   --> $DIR/generics.rs:14:52
+   --> $DIR/generics.rs:15:52
9    |
10 LL | struct Test<T: Copy> {
11    |             - similarly named type parameter `T` defined here

22    |                    +++
23 
24 error[E0562]: `impl Trait` is not allowed in `fn` pointer parameters
+   --> $DIR/generics.rs:17:43
26    |
26    |
27 LL |     f2: extern "C-cmse-nonsecure-call" fn(impl Copy, u32, u32, u32) -> u64,

30    = note: `impl Trait` is only allowed in arguments and return types of functions and methods
31 
31 
32 error[E0798]: function pointers with the `"C-cmse-nonsecure-call"` ABI cannot contain generics in their type
+   --> $DIR/generics.rs:19:43
34    |
34    |
35 LL |     f3: extern "C-cmse-nonsecure-call" fn(T, u32, u32, u32) -> u64,

37 
37 
38 error[E0798]: function pointers with the `"C-cmse-nonsecure-call"` ABI cannot contain generics in their type
+   --> $DIR/generics.rs:20:43
40    |
40    |
41 LL |     f4: extern "C-cmse-nonsecure-call" fn(Wrapper<T>, u32, u32, u32) -> u64,


The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/cmse-nonsecure/cmse-nonsecure-call/generics/generics.stderr
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/cmse-nonsecure/cmse-nonsecure-call/generics/generics.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args cmse-nonsecure/cmse-nonsecure-call/generics.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/generics.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/cmse-nonsecure/cmse-nonsecure-call/generics" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/cmse-nonsecure/cmse-nonsecure-call/generics/auxiliary" "--target" "thumbv8m.main-none-eabi" "--crate-type" "lib"
--- stderr -------------------------------
error: function pointer types may not have generic parameters
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/generics.rs:15:42
   |
   |
LL |     f1: extern "C-cmse-nonsecure-call" fn<U: Copy>(U, u32, u32, u32) -> u64, //~ ERROR cannot find type `U` in this scope

error[E0412]: cannot find type `U` in this scope
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/generics.rs:15:52
   |
   |
LL | struct Test<T: Copy> {
   |             - similarly named type parameter `T` defined here
LL |     f1: extern "C-cmse-nonsecure-call" fn<U: Copy>(U, u32, u32, u32) -> u64, //~ ERROR cannot find type `U` in this scope
   |
help: a type parameter with a similar name exists
   |
   |
LL |     f1: extern "C-cmse-nonsecure-call" fn<U: Copy>(T, u32, u32, u32) -> u64, //~ ERROR cannot find type `U` in this scope
help: you might be missing a type parameter
   |
   |
LL | struct Test<T: Copy, U> {


error[E0562]: `impl Trait` is not allowed in `fn` pointer parameters
   |
   |
LL |     f2: extern "C-cmse-nonsecure-call" fn(impl Copy, u32, u32, u32) -> u64,
   |
   = note: `impl Trait` is only allowed in arguments and return types of functions and methods


error[E0798]: function pointers with the `"C-cmse-nonsecure-call"` ABI cannot contain generics in their type
   |
   |
LL |     f3: extern "C-cmse-nonsecure-call" fn(T, u32, u32, u32) -> u64, //~ ERROR [E0798]


error[E0798]: function pointers with the `"C-cmse-nonsecure-call"` ABI cannot contain generics in their type
   |
   |
LL |     f4: extern "C-cmse-nonsecure-call" fn(Wrapper<T>, u32, u32, u32) -> u64, //~ ERROR [E0798]

error: aborting due to 5 previous errors

Some errors have detailed explanations: E0412, E0562, E0798.
Some errors have detailed explanations: E0412, E0562, E0798.
For more information about an error, try `rustc --explain E0412`.
------------------------------------------


---- [ui] tests/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack.rs stdout ----
diff of stderr:

1 error[E0798]: arguments for `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/params-via-stack.rs:16:43
+   --> $DIR/params-via-stack.rs:17:43
3    |
4 LL |     f1: extern "C-cmse-nonsecure-call" fn(u32, u32, u32, u32, u32),
5    |                                           ^^^^^^^^^^^^^^^^^^^^^^^ these arguments don't fit in the available registers

7    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass all their arguments via the 4 32-bit available argument registers
8 
9 error[E0798]: arguments for `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/params-via-stack.rs:17:43
+   --> $DIR/params-via-stack.rs:18:43
11    |
12 LL |     f2: extern "C-cmse-nonsecure-call" fn(u32, u32, u32, u16, u16),
13    |                                           ^^^^^^^^^^^^^^^^^^^^^^^ these arguments don't fit in the available registers

15    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass all their arguments via the 4 32-bit available argument registers
16 
17 error[E0798]: arguments for `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/params-via-stack.rs:18:43
+   --> $DIR/params-via-stack.rs:19:43
19    |
20 LL |     f3: extern "C-cmse-nonsecure-call" fn(u32, u64, u32),
21    |                                           ^^^^^^^^^^^^^ these arguments don't fit in the available registers

23    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass all their arguments via the 4 32-bit available argument registers
24 
25 error[E0798]: arguments for `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/params-via-stack.rs:19:43
+   --> $DIR/params-via-stack.rs:20:43
27    |
28 LL |     f4: extern "C-cmse-nonsecure-call" fn(AlignRelevant, u32),
29    |                                           ^^^^^^^^^^^^^^^^^^ these arguments don't fit in the available registers

31    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass all their arguments via the 4 32-bit available argument registers
32 
33 error[E0798]: arguments for `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/params-via-stack.rs:20:43
+   --> $DIR/params-via-stack.rs:21:43
35    |
36 LL |     f5: extern "C-cmse-nonsecure-call" fn([u32; 5]),
37    |                                           ^^^^^^^^ these arguments don't fit in the available registers

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack/params-via-stack.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args cmse-nonsecure/cmse-nonsecure-call/params-via-stack.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack/auxiliary" "--target" "thumbv8m.main-none-eabi" "--crate-type" "lib"
--- stderr -------------------------------
--- stderr -------------------------------
error[E0798]: arguments for `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack.rs:17:43
   |
LL |     f1: extern "C-cmse-nonsecure-call" fn(u32, u32, u32, u32, u32), //~ ERROR [E0798]
   |                                           ^^^^^^^^^^^^^^^^^^^^^^^ these arguments don't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass all their arguments via the 4 32-bit available argument registers

error[E0798]: arguments for `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack.rs:18:43
   |
LL |     f2: extern "C-cmse-nonsecure-call" fn(u32, u32, u32, u16, u16), //~ ERROR [E0798]
   |                                           ^^^^^^^^^^^^^^^^^^^^^^^ these arguments don't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass all their arguments via the 4 32-bit available argument registers

error[E0798]: arguments for `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack.rs:19:43
   |
LL |     f3: extern "C-cmse-nonsecure-call" fn(u32, u64, u32),           //~ ERROR [E0798]
   |                                           ^^^^^^^^^^^^^ these arguments don't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass all their arguments via the 4 32-bit available argument registers

error[E0798]: arguments for `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack.rs:20:43
   |
LL |     f4: extern "C-cmse-nonsecure-call" fn(AlignRelevant, u32),      //~ ERROR [E0798]
   |                                           ^^^^^^^^^^^^^^^^^^ these arguments don't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass all their arguments via the 4 32-bit available argument registers

error[E0798]: arguments for `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/params-via-stack.rs:21:43
   |
LL |     f5: extern "C-cmse-nonsecure-call" fn([u32; 5]),                //~ ERROR [E0798]
   |                                           ^^^^^^^^ these arguments don't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass all their arguments via the 4 32-bit available argument registers
error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0798`.
------------------------------------------
------------------------------------------


---- [ui] tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs stdout ----
diff of stderr:

1 error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/return-via-stack.rs:34:50
+   --> $DIR/return-via-stack.rs:35:50
3    |
4 LL |     u128: extern "C-cmse-nonsecure-call" fn() -> u128,
5    |                                                  ^^^^ this type doesn't fit in the available registers

7    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers
8 
9 error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/return-via-stack.rs:35:50
+   --> $DIR/return-via-stack.rs:36:50
11    |
12 LL |     i128: extern "C-cmse-nonsecure-call" fn() -> i128,
13    |                                                  ^^^^ this type doesn't fit in the available registers

15    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers
16 
17 error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/return-via-stack.rs:24:48
+   --> $DIR/return-via-stack.rs:25:48
19    |
20 LL |     f1: extern "C-cmse-nonsecure-call" fn() -> ReprCU64,
21    |                                                ^^^^^^^^ this type doesn't fit in the available registers

23    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers
24 
25 error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/return-via-stack.rs:25:48
+   --> $DIR/return-via-stack.rs:26:48
27    |
28 LL |     f2: extern "C-cmse-nonsecure-call" fn() -> ReprCBytes,
29    |                                                ^^^^^^^^^^ this type doesn't fit in the available registers

31    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers
32 
33 error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/return-via-stack.rs:26:48
+   --> $DIR/return-via-stack.rs:27:48
35    |
36 LL |     f3: extern "C-cmse-nonsecure-call" fn() -> U64Compound,
37    |                                                ^^^^^^^^^^^ this type doesn't fit in the available registers

39    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers
40 
41 error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/return-via-stack.rs:27:48
+   --> $DIR/return-via-stack.rs:28:48
43    |
44 LL |     f4: extern "C-cmse-nonsecure-call" fn() -> ReprCAlign16,
45    |                                                ^^^^^^^^^^^^ this type doesn't fit in the available registers

47    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers
48 
49 error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/return-via-stack.rs:28:48
+   --> $DIR/return-via-stack.rs:29:48
51    |
52 LL |     f5: extern "C-cmse-nonsecure-call" fn() -> [u8; 5],
53    |                                                ^^^^^^^ this type doesn't fit in the available registers

55    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers
56 
57 error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/return-via-stack.rs:50:48
+   --> $DIR/return-via-stack.rs:51:48
59    |
60 LL |     f1: extern "C-cmse-nonsecure-call" fn() -> ReprRustUnionU64,
61    |                                                ^^^^^^^^^^^^^^^^ this type doesn't fit in the available registers

63    = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers
64 
65 error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
-   --> $DIR/return-via-stack.rs:51:48
+   --> $DIR/return-via-stack.rs:52:48
67    |
68 LL |     f2: extern "C-cmse-nonsecure-call" fn() -> ReprCUnionU64,
69    |                                                ^^^^^^^^^^^^^ this type doesn't fit in the available registers

The actual stderr differed from the expected stderr.
Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack/return-via-stack.stderr
To update references, rerun the tests and pass the `--bless` flag
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack/auxiliary" "--target" "thumbv8m.main-none-eabi" "--crate-type" "lib"
--- stderr -------------------------------
--- stderr -------------------------------
error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs:35:50
   |
LL |     u128: extern "C-cmse-nonsecure-call" fn() -> u128, //~ ERROR [E0798]
   |                                                  ^^^^ this type doesn't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers

error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs:36:50
   |
LL |     i128: extern "C-cmse-nonsecure-call" fn() -> i128, //~ ERROR [E0798]
   |                                                  ^^^^ this type doesn't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers

error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs:25:48
   |
LL |     f1: extern "C-cmse-nonsecure-call" fn() -> ReprCU64, //~ ERROR [E0798]
   |                                                ^^^^^^^^ this type doesn't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers

error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs:26:48
   |
LL |     f2: extern "C-cmse-nonsecure-call" fn() -> ReprCBytes, //~ ERROR [E0798]
   |                                                ^^^^^^^^^^ this type doesn't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers

error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs:27:48
   |
LL |     f3: extern "C-cmse-nonsecure-call" fn() -> U64Compound, //~ ERROR [E0798]
   |                                                ^^^^^^^^^^^ this type doesn't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers

error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs:28:48
   |
LL |     f4: extern "C-cmse-nonsecure-call" fn() -> ReprCAlign16, //~ ERROR [E0798]
   |                                                ^^^^^^^^^^^^ this type doesn't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers

error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs:29:48
   |
LL |     f5: extern "C-cmse-nonsecure-call" fn() -> [u8; 5],  //~ ERROR [E0798]
   |                                                ^^^^^^^ this type doesn't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers

error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs:51:48
   |
LL |     f1: extern "C-cmse-nonsecure-call" fn() -> ReprRustUnionU64, //~ ERROR [E0798]
   |                                                ^^^^^^^^^^^^^^^^ this type doesn't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers

error[E0798]: return value of `"C-cmse-nonsecure-call"` function too large to pass via registers
##[error]  --> /checkout/tests/ui/cmse-nonsecure/cmse-nonsecure-call/return-via-stack.rs:52:48
   |
LL |     f2: extern "C-cmse-nonsecure-call" fn() -> ReprCUnionU64,    //~ ERROR [E0798]
   |                                                ^^^^^^^^^^^^^ this type doesn't fit in the available registers
   |
   = note: functions with the `"C-cmse-nonsecure-call"` ABI must pass their result via the available return registers
error: aborting due to 9 previous errors

For more information about this error, try `rustc --explain E0798`.
------------------------------------------
```

</details>

oli-obk commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs:
```diff
@@ -0,0 +1,152 @@
+        // fn(u32, u32, u32, u16, u16) -> u32,
+        //    ^^^^^^^^^^^^^^^^^^^^^^^     ^^^
+        let output_span = bare_fn_ty.decl.output.span();
+        let inputs_span = match (
```
This could even be used to add some information about each input's size so the user doesn't have to re-do the math in their head?

folkertdev replied:
the error message now underlines the arguments that are too large. We believe that, because this is a very niche feature that requires significant knowledge of the underlying hardware, users should be able to figure it out based on this error. If not, the error markdown has more information and links.

folkertdev committed:
improve error reporting

folkertdev commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs:
```diff
@@ -0,0 +1,152 @@
+    fn_sig: ty::PolyFnSig<'tcx>,
+) {
+    if let abi::Abi::CCmseNonSecureCall = abi {
+        let hir_node = tcx.hir_node(hir_id);
+        let hir::Node::Ty(hir::Ty {
+            span: bare_fn_span,
+            kind: hir::TyKind::BareFn(bare_fn_ty),
+            ..
+        }) = hir_node
+        else {
+            // might happen when this ABI is used incorrectly. That will be handled elsewhere
+            return;
+        };
+
```
`decl` does not have sufficient span information. In particular it lacks the span of identifiers in named parameters `x: u32`.

I too think fetching the hir_node again is inelegant, but passing all of that information down seems impractical. We only do it for function pointers with this very niche ABI, so performance-wise I don't see a problem.

folkertdev commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs:
```diff
@@ -0,0 +1,152 @@
+    if size <= 4 {
+        return Ok(true);
+    } else if size > 8 {
+        return Ok(false);
```
we've clarified in the error message what the reasons for rejection are

folkertdev commented:
@rustbot ready

oli-obk commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/cmse.rs:
```diff
@@ -72,21 +68,29 @@ pub fn validate_cmse_abi<'tcx>(
 
-    for arg_def in fn_sig.inputs().iter() {
+    for (index, arg_def) in fn_sig.inputs().iter().enumerate() {
         let layout = tcx.layout_of(ParamEnv::reveal_all().and(*arg_def.skip_binder()))?;
```
I think at this point it may be easier to just emit the errors in `is_valid_cmse_inputs` and `_output` directly. Then you can also report a "generics not allowed" error with a span pointing at the offending thing.

folkertdev replied:
we have 2 reasons not to do that

* `arg_def` doesn't have any span information. it's effectively just a `Ty`
* in the (near, hopefully) future, the `C-cmse-nonsecure-entry` abi will require the same validation, and we'd like to reuse this code, so passing in `BareFn`-specific data will not work

oli-obk commented:
@bors r+ rollup

bors commented:
:pushpin: Commit c2894a4297fd1e824e071fd21a92bfe475c0169a has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk c2894a4297fd1e824e071fd21a92bfe475c0169a -->
<!-- homu: {"type":"Approved","sha":"c2894a4297fd1e824e071fd21a92bfe475c0169a","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
